clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

```diff
@@ -4102,6 +4102,238 @@ void collect_unreachable(mlir::Operation *parent,
   }
 }
 
+static void buildGlobalAnnotationsVar(mlir::ModuleOp module) {
+  auto annotationValuesAttr = module->getAttr("cir.global_annotations");
+  StringRef annotationSection = "llvm.metadata";
+  if (annotationValuesAttr) {
+    auto annotationValuesArray =
+        mlir::cast<mlir::ArrayAttr>(annotationValuesAttr);
+    if (!annotationValuesArray || annotationValuesArray.empty())
+      return;
+    mlir::OpBuilder globalVarBuilder(module.getContext());
+    globalVarBuilder.setInsertionPointToEnd(&module.getBodyRegion().front());
+
+    // Create a global array for annotation values with element type of
+    // struct { ptr, ptr, ptr, i32, ptr }
+    auto annoPtrTy =
+        mlir::LLVM::LLVMPointerType::get(globalVarBuilder.getContext());
+    llvm::SmallVector<mlir::Type> annoStructFields;
+    annoStructFields.push_back(annoPtrTy);
+    annoStructFields.push_back(annoPtrTy);
+    annoStructFields.push_back(annoPtrTy);
+    annoStructFields.push_back(globalVarBuilder.getI32Type());
+    annoStructFields.push_back(annoPtrTy);
+
+    auto annoStructTy = mlir::LLVM::LLVMStructType::getLiteral(
+        globalVarBuilder.getContext(), annoStructFields);
+    auto annoStructArrayTy = mlir::LLVM::LLVMArrayType::get(
+        annoStructTy, annotationValuesArray.size());
+
+    auto loc = module.getLoc();
+    auto annotationGlobalOp = globalVarBuilder.create<mlir::LLVM::GlobalOp>(
+        loc, annoStructArrayTy, false, mlir::LLVM::Linkage::Appending,
+        "llvm.global.annotations", mlir::Attribute());
+    annotationGlobalOp.setSection("llvm.metadata");
+    annotationGlobalOp.getRegion().push_back(new mlir::Block());
+    mlir::OpBuilder varInitBuilder(module.getContext());
+    varInitBuilder.setInsertionPointToEnd(
+        annotationGlobalOp.getInitializerBlock());
+    // Globals created for annotation strings and args to be
+    // placed before the var llvm.global.annotations. This is consistent
+    // with clang code gen.
+    globalVarBuilder.setInsertionPoint(annotationGlobalOp);
+
+    auto lowerAnnotationValuesArray = [&](mlir::ArrayAttr annotationValues) {
```

Comment:
The content of this function isn't how we write code in LLVM in general. Lambdas are supposed to be used for short snippet of code, and almost never nested. Please break this into as many functions as possible to make this readable and easy to follow.

Also note that it doesn't need to be necessarily a static function, it could one or more methods of `ConvertCIRToLLVMPass`.